### PR TITLE
Add github signin / signup function

### PR DESCRIPTION
**Pull request by richardartoul**
Right now its a separate function because we don't have it setup everywhere properly yet, but once we do have it set up everywhere we'll change it to integrate with the login/signup commands automatically

### CODE REVIEW

File: hyperdash_cli/cli.py [hunk 1]
```diff
@@ -128,6 +134,85 @@ def demo(args=None):
...
+    # a race condition where the Hyperdash server redirects with an access token but
+    # the Python server isn't read yet
+    print("Opening browser, please wait. If something goes wrong, press CTRL+C to cancel.")
+    server_started_queue.get(block=True)
+    # Blocks until browser opens, but doesn't wait for user to close it
+    # TODO: Constantize this
+    webbrowser.open_new_tab("https://github.com/login/oauth/authorize?client_id=4793c654216adde7a478&redirect_uri=http://localhost:4040/oauth/github/callback&allow_signup=false&scope=user%20email&state=client_cli:{}".format(port))
+
+
+    print("Waiting for Github OAuth to complete. If something goes wrong, press CTRL+C to cancel.")
```
Comment (spence) [inline hyperdash_cli/cli.py]: 2 lines?

File: hyperdash_cli/cli.py [hunk 2]
```diff
@@ -128,6 +134,85 @@ def demo(args=None):
...
+    def start_server():
+        class OAuthRedirectHandler(BaseHTTPServer.BaseHTTPRequestHandler):
+            def do_GET(self):
+                parsed_path = urlparse(self.path)
+                query = parse_qs(parsed_path.query)
+                access_token = query["access_token"][0]
+                access_token_queue.put(access_token)
+                # Redirect user's browser
+                self.send_response(301)
+                self.send_header('Location',"{}/{}".format(get_base_http_url(), "/oauth/github/success"))
```
Comment (spence) [inline hyperdash_cli/cli.py]: Use consistent quotes.

File: hyperdash_cli/cli.py [hunk 3]
```diff
@@ -128,6 +134,85 @@ def demo(args=None):
...
+    access_token_queue = Queue()
+
+    # Server that we will run in the background to accept a post-OAuth redirect from
+    # the Hyperdash server which will contain the user's access token
+    def start_server():
+        class OAuthRedirectHandler(BaseHTTPServer.BaseHTTPRequestHandler):
+            def do_GET(self):
+                parsed_path = urlparse(self.path)
+                query = parse_qs(parsed_path.query)
+                access_token = query["access_token"][0]
```
Comment (spence) [inline hyperdash_cli/cli.py]: Maybe 
```python
query["access_token"][0] if "access_token" in query else None
```
and handle `None`

File: hyperdash_cli/cli.py [hunk 4]
```diff
@@ -128,6 +134,85 @@ def demo(args=None):
...
+                access_token_queue.put(access_token)
+                # Redirect user's browser
+                self.send_response(301)
+                self.send_header('Location',"{}/{}".format(get_base_http_url(), "/oauth/github/success"))
+                self.end_headers()
+
+        server = BaseHTTPServer.HTTPServer(('localhost', port), OAuthRedirectHandler)
```
Comment (spence) [inline hyperdash_cli/cli.py]: Might want to bind to `127.0.0.1` or `0.0.0.0`.

File: hyperdash_cli/cli.py [hunk 5]
```diff
@@ -128,6 +134,85 @@ def demo(args=None):
...
+    server_thread.setDaemon(True)
+    server_thread.start()
+    # Wait until the server has started before opening the user's browser to prevent
+    # a race condition where the Hyperdash server redirects with an access token but
+    # the Python server isn't read yet
+    print("Opening browser, please wait. If something goes wrong, press CTRL+C to cancel.")
+    server_started_queue.get(block=True)
+    # Blocks until browser opens, but doesn't wait for user to close it
+    # TODO: Constantize this
+    webbrowser.open_new_tab("https://github.com/login/oauth/authorize?client_id=4793c654216adde7a478&redirect_uri=http://localhost:4040/oauth/github/callback&allow_signup=false&scope=user%20email&state=client_cli:{}".format(port))
```
Comment (spence) [inline hyperdash_cli/cli.py]: You have `4040` hardcoded as the local port in this.

File: hyperdash_cli/cli.py [hunk 6]
```diff
@@ -128,6 +134,85 @@ def demo(args=None):
...
+    success, default_api_key = _after_access_token_login(access_token)
+    if success:
+        print("Successfully logged in! We also installed: {} as your default API key".format(
+            default_api_key))
+
+
+def _find_available_port():
+    port = None
+    for cur_port in xrange(3000, 3100):
+        is_open = _is_port_open("127.0.0.1", cur_port)
```
Comment (spence) [inline hyperdash_cli/cli.py]: For kicks, increase search range?

File: hyperdash_cli/cli.py [hunk 7]
```diff
@@ -128,6 +134,85 @@ def demo(args=None):
...
+        if is_open:
+            port = cur_port
+            break
+    return port
```
Comment (spence) [inline hyperdash_cli/cli.py]: Can simply code:
```python
for cur_port in xrange(3000, 3100):
  if _is_port_open("127.0.0.1", cur_port):
      return cur_port
return None
```

File: hyperdash_cli/cli.py [hunk 8]
```diff
@@ -128,6 +134,85 @@ def demo(args=None):
...
+
+
+def _is_port_open(host, port):
+    try:
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.bind(('127.0.0.1', port))
+        sock.listen(5)
+        sock.close()
+        sock = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
+        sock.bind(('::1', port))
```
Comment (spence) [inline hyperdash_cli/cli.py]: Are we using single or double quotes.

File: hyperdash_cli/cli.py [hunk 9]
```diff
@@ -128,6 +134,85 @@ def demo(args=None):
...
+    return port
+
+
+def _is_port_open(host, port):
+    try:
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.bind(('127.0.0.1', port))
+        sock.listen(5)
+        sock.close()
+        sock = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
```
Comment (spence) [inline hyperdash_cli/cli.py]: Skip ipv6 test if not supported.

https://docs.python.org/2/library/socket.html#socket.has_ipv6

File: hyperdash_cli/cli.py [hunk 10]
```diff
@@ -128,6 +134,85 @@ def demo(args=None):
...
+
+    print("Waiting for Github OAuth to complete. If something goes wrong, press CTRL+C to cancel.")
+    # Wait for the Hyperdash server to redirect with the access token or an error
+    access_token = access_token_queue.get(block=True)
+    # Use the access token to retrieve the user's API key and store a valid
+    # hyperdash.json file
+    success, default_api_key = _after_access_token_login(access_token)
+    if success:
+        print("Successfully logged in! We also installed: {} as your default API key".format(
+            default_api_key))
```
Comment (spence) [inline hyperdash_cli/cli.py]:
```python
else:
   # TODO
```

**Comment by spence** [comment]
Really think we should use `0.0.0.0` for all hostnames in this. It's more robust in its meaning and less likely to be subject to local user environment or changes.

File: hyperdash_cli/cli.py [hunk 11]
```diff
@@ -128,6 +134,85 @@ def demo(args=None):
...
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        sock.bind(('127.0.0.1', port))
+        sock.listen(5)
+        sock.close()
+        sock = socket.socket(socket.AF_INET6, socket.SOCK_STREAM)
+        sock.bind(('::1', port))
+        sock.listen(5)
+        sock.close()
+    except socket.error as e:
+        return False
```
Comment (andrewschreiber) [inline hyperdash_cli/cli.py]: Looked into error handling for `socket`. Turns out there are a few exception types beyond `socket.error`. Doesn't seem like we're probable to trigger the others with the `_is_port_open` logic, but I'd recommend a more generic error catcher to be safe.

> https://docs.python.org/2/library/socket.html
> exception socket.error
> This exception is raised for socket-related errors. The accompanying value is either a string telling what went wrong or a pair (errno, string) representing an error returned by a system call, similar to the value accompanying os.error. See the module errno, which contains names for the error codes defined by the underlying operating system.
> 
> exception socket.herror
> This exception is raised for address-related errors, i.e. for functions that use h_errno in the C API, including gethostbyname_ex() and gethostbyaddr().
> 
> The accompanying value is a pair (h_errno, string) representing an error returned by a library call. string represents the description of h_errno, as returned by the hstrerror() C function.
> 
> exception socket.gaierror
> This exception is raised for address-related errors, for getaddrinfo() and getnameinfo(). The accompanying value is a pair (error, string) representing an error returned by a library call. string represents the description of error, as returned by the gai_strerror() C function. The error value will match one of the EAI_* constants defined in this module.
> 
> exception socket.timeout
> This exception is raised when a timeout occurs on a socket which has had timeouts enabled via a prior call to settimeout(). The accompanying value is a string whose value is currently always “timed out”.

File: hyperdash_cli/cli.py [hunk 12]
```diff
@@ -128,6 +134,85 @@ def demo(args=None):
...
+        server_started_queue.put(True)
+        server.handle_request()
+    
+    server_thread = Thread(target=start_server)
+    # Prevent server_thread from preventing program shutdown
+    server_thread.setDaemon(True)
+    server_thread.start()
+    # Wait until the server has started before opening the user's browser to prevent
+    # a race condition where the Hyperdash server redirects with an access token but
+    # the Python server isn't read yet
```
Comment (andrewschreiber) [inline hyperdash_cli/cli.py]: Good call.

File: hyperdash_cli/cli.py (hunk 2 above)
Comment (andrewschreiber) [reply hyperdash_cli/cli.py]: A linter is the systematic answer here. 

@richardartoul @spence Do you guys have a preference for a python linter? My constraint is that I want VSCode's python plugin to support it, leaving us with these three options:
https://www.pylint.org/
https://pypi.python.org/pypi/pep8
https://flake8.readthedocs.io/en/latest/

File: hyperdash_cli/cli.py (hunk 10 above)
Comment (richardartoul) [reply hyperdash_cli/cli.py]: the _after_access_token_login already handles it

File: hyperdash_cli/cli.py (hunk 11 above)
Comment (richardartoul) [reply hyperdash_cli/cli.py]: for socket.herror we're not using any of the address-related functions, same for socket.gaierror. Good call on the timeout one, I meant to add a timeout but forgot

File: hyperdash_cli/constants.py [hunk 13]
```diff
@@ -3,6 +3,22 @@
 import six
 
 
+GITHUB_CLIENT_ID = "4793c654216adde7a478"
+GITHUB_OAUTH_SCOPES = ["user:email"]
+GITHUB_REDIRECT_URI_PATH = "oauth/github/callback"
```
Comment (spence) [inline hyperdash_cli/constants.py]: hd redirect?

**Pull request merged**